src/Compilers/Core/Portable/Compilation/IExpression.cs

```diff
@@ -0,0 +1,717 @@
+        /// and params/ParamArray arguments have been collected into arrays. Arguments are not present
+        /// unless supplied in source.
+        /// </summary>
+        ImmutableArray<IArgument> ArgumentsInSourceOrder { get; }
+        /// <summary>
+        /// Find the argument supplied for a given parameter of the target method.
+        /// </summary>
+        /// <param name="parameter">Parameter of the target method.</param>
+        /// <returns>Argument corresponding to the parameter.</returns>
+        IArgument ArgumentMatchingParameter(IParameterSymbol parameter);
+    }
+
+    /// <summary>
+    /// Represents an argument in a method invocation.
+    /// </summary>
+    public interface IArgument
+    {
+        /// <summary>
+        /// Kind of argument.
+        /// </summary>
+        ArgumentKind Kind { get; }
+        /// <summary>
+        /// Parameter the argument matches.
+        /// </summary>
+        IParameterSymbol Parameter { get; }
+        /// <summary>
+        /// Value supplied for the argument.
+        /// </summary>
+        IExpression Value { get; }
+        /// <summary>
+        /// Conversion applied to the argument value passing it into the target method. Applicable only to VB Reference arguments.
+        /// </summary>
+        IExpression InConversion { get; }
+        /// <summary>
+        /// Conversion applied to the argument value after the invocation. Applicable only to VB Reference arguments.
+        /// </summary>
+        IExpression OutConversion { get; }
+    }
+
+    /// <summary>
+    /// Kinds of arguments.
+    /// </summary>
+    public enum ArgumentKind
+    {
+        /// <summary>
+        /// Argument is specified positionally and matches the parameter of the same ordinality.
+        /// </summary>
+        Positional,
+        /// <summary>
+        /// Argument is specified by name and matches the parameter of the same name.
+        /// </summary>
+        Named,
+        /// <summary>
+        /// Argument becomes an element of an array that matches a trailing C# params or VB ParamArray parameter.
+        /// </summary>
+        ParamArray,
+        /// <summary>
+        /// Argument was omitted in source but has a default value supplied automatically.
+        /// </summary>
+        DefaultValue
+    }
+
+    /// <summary>
+    /// Represents a reference, which refers to a symbol or an element of a collection.
+    /// </summary>
+    public interface IReferenceExpression : IExpression
+    {
+        /// <summary>
+        /// Kind of the reference.
+        /// </summary>
+        ReferenceKind ReferenceKind { get; }
+    }
+
+    /// <summary>
+    /// Kinds of references.
+    /// </summary>
+    public enum ReferenceKind
+    {
+        Local,
+        Parameter,
+        SyntheticLocal,
+        ArrayElement,
+        PointerIndirection,
+        StaticField,
+        InstanceField,
```

Comment:
Yes, but I'm not sure this level of redundancy is a bad thing. Maybe it leads to analyzers having to test for two values (e.g. StaticField and InstanceField), in which case it would be a bad thing. I'll have a more coherent opinion after sleeping.